src/normalize.ts

```diff
@@ -0,0 +1,31 @@
+import { format } from './format';
+import { parse } from './parse';
+import { IPath } from './types';
+
+export function normalize(filepath: string) {
+  return format(normalizeParsed(parse(filepath)));
+}
+
+export function normalizeParsed(parsed: IPath): IPath {
+  let path = parsed.path;
+
+  // Replace consecutive '/' and replace '/./' with '/'
+  path = path.filter(segment => segment !== '' && segment !== '.');
+
+  // Collapse '..' where possible
+  const stack = [];
+  for (const segment of path) {
+    if (segment === '..') {
+      if (stack.length && stack[stack.length - 1] !== '..') {
```

Comment:
The flow here could be simplified (more readable) a tiny bit.
```ts
if (stack.length === 0) continue;
if (stack[stack.length - 1] !== '..') {
  stack.pop();
} else if (!parsed.absolute) {
  stack.push(segment);
}
```

Reply:
:D Actually that implementation is wrong. But that demonstrates your core argument (the logic is too complicated to follow) is correct. I'll see what I can do.

Reply:
> Actually that implementation is wrong. 

How so? I didn't mean to replace everything in the loop, just this particular block